src/main/asciidoc/docs/testng_listeners.adoc

```diff
@@ -214,3 +214,142 @@ Finishing
+* Now this implementation needs to be plugged into TestNG via the configuration parameter `-listenercomparator`
+
+CAUTION: TestNG orders only user's listeners and which are not part of the exclusion list that can be specified via the JVM argument `-Dtestng.preferential.listeners.package` (Multiple fully qualified class names can be specified as comma separated values). If nothing is specified, TestNG by default excludes all the IntelliJ IDEA listeners under the package `com.intellij.rt.*`.
+
+Let's see an example.
+
+* Let's create a custom annotation that will help us define the order in which *our* listeners should be invoked.
+
+[source, java]
+
+----
+@Retention(java.lang.annotation.RetentionPolicy.RUNTIME)
+@Target({ TYPE})
+
+public @interface RunOrder {
+  int value();
+}
+----
+
+* Now let's create an implementation of `org.testng.ListenerComparator` which honours this custom annotation that we just now created.
+
+[source, java]
+
+----
+import org.testng.ITestNGListener;
+import org.testng.ListenerComparator;
+
+public class AnnotationBackedListenerComparator implements ListenerComparator {
+
+  @Override
+  public int compare(ITestNGListener l1, ITestNGListener l2) {
+    RunOrder one = l1.getClass().getAnnotation(RunOrder.class);
+    RunOrder two = l2.getClass().getAnnotation(RunOrder.class);
+    if (one != null && two != null) {
+      return Integer.compare(one.value(), two.value());
+    }
+    //Add all other listeners which don't have the @RunOrder annotation to the end so that they get executed
+    //at the last after all the ordered listeners are executed.
+    return -1;
+  }
+}
+----
+
+* Lets say we have the below listeners as samples.
+
+[source, java]
+
+----
+import org.testng.IExecutionListener;
+
+public class ExecutionListenerHolder {
+
+  public static final String PREFIX = ExecutionListenerHolder.class.getName() + "$";
+
+  public abstract static class KungFuWarrior implements IExecutionListener {
+    @Override
+    public void onExecutionStart() {
+      System.err.println(getClass().getSimpleName() + ".onExecutionStart");
+    }
+
+    @Override
+    public void onExecutionFinish() {
+      System.err.println(getClass().getSimpleName() + ".onExecutionFinish");
+    }
+  }
+
+  @RunOrder(1)
+  public static class MasterOogway extends KungFuWarrior { }
+
+  @RunOrder(2)
+  public static class MasterShifu extends KungFuWarrior { }
+
+  public static class DragonWarrior extends KungFuWarrior { }
+}
+----
+
+* A sample code snippet that uses the TestNG APIs to run the test could look like below:
+
+[source, java]
+
+----
+TestNG testng = create(NormalSampleTestCase.class);
+testng.addListener(new ExecutionListenerHolder.DragonWarrior());
+testng.addListener(new ExecutionListenerHolder.MasterShifu());
+testng.addListener(new ExecutionListenerHolder.MasterOogway());
+testng.setListenerComparator(new AnnotationBackedListenerComparator());
```

Comment:
`testng.setListenerComparator(AnnotationBackedListenerComparator.class);` instead

Reply:
The TestNG API always was meant to be accepting an instance because this value can be set either via this method or via the configuration parameter which accepts a fully qualified class name. I chose this because I didn't see a lot of value add in terms of allowing users to set the fully qualified class name when using the TestNG API.

Reply:
Ok